planning/autoware_path_generator/src/utils.cpp

```diff
@@ -0,0 +1,212 @@
+  const double backward_distance)
+{
+  if (!exists(planner_data.route_lanelets, lanelet)) {
+    return std::nullopt;
+  }
+
+  const auto arc_coordinates = lanelet::utils::getArcCoordinates({lanelet}, current_pose);
+  const auto lanelet_length = lanelet::utils::getLaneletLength2d(lanelet);
+
+  const auto backward_lanelets =
+    get_lanelets_up_to(lanelet, planner_data, backward_distance - arc_coordinates.length);
+  if (!backward_lanelets) {
+    return std::nullopt;
+  }
+
+  const auto forward_lanelets = get_lanelets_after(
+    lanelet, planner_data, forward_distance - (lanelet_length - arc_coordinates.length));
+  if (!forward_lanelets) {
+    return std::nullopt;
+  }
+
+  lanelet::ConstLanelets lanelets(std::move(*backward_lanelets));
+  lanelets.push_back(lanelet);
+  std::move(forward_lanelets->begin(), forward_lanelets->end(), std::back_inserter(lanelets));
+
+  return lanelets;
+}
+
+std::optional<lanelet::ConstLanelets> get_lanelets_after(
+  const lanelet::ConstLanelet & lanelet, const PlannerData & planner_data, const double distance)
+{
+  if (!exists(planner_data.route_lanelets, lanelet)) {
+    return std::nullopt;
+  }
+
+  lanelet::ConstLanelets lanelets{};
+  auto current_lanelet = lanelet;
+  auto length = 0.;
+
+  while (rclcpp::ok() && length < distance) {
+    const auto next_lanelet = get_next_lanelet_within_route(current_lanelet, planner_data);
+    if (!next_lanelet) {
+      break;
+    }
+
+    lanelets.push_back(*next_lanelet);
+    current_lanelet = *next_lanelet;
+    length += lanelet::utils::getLaneletLength2d(*next_lanelet);
+  }
+
+  return lanelets;
+}
+
+std::optional<lanelet::ConstLanelets> get_lanelets_up_to(
+  const lanelet::ConstLanelet & lanelet, const PlannerData & planner_data, const double distance)
+{
+  if (!exists(planner_data.route_lanelets, lanelet)) {
+    return std::nullopt;
+  }
+
+  lanelet::ConstLanelets lanelets{};
+  auto current_lanelet = lanelet;
+  auto length = 0.;
+
+  while (rclcpp::ok() && length < distance) {
+    const auto prev_lanelet = get_previous_lanelet_within_route(current_lanelet, planner_data);
+    if (!prev_lanelet) {
+      break;
+    }
+
+    lanelets.push_back(*prev_lanelet);
+    current_lanelet = *prev_lanelet;
+    length += lanelet::utils::getLaneletLength2d(*prev_lanelet);
+  }
+
+  std::reverse(lanelets.begin(), lanelets.end());
+  return lanelets;
+}
+
+std::optional<lanelet::ConstLanelet> get_next_lanelet_within_route(
+  const lanelet::ConstLanelet & lanelet, const PlannerData & planner_data)
+{
+  if (planner_data.preferred_lanelets.empty()) {
+    return std::nullopt;
+  }
+
+  if (exists(planner_data.goal_lanelets, lanelet)) {
+    return std::nullopt;
+  }
+
+  for (const auto & lanelet : planner_data.routing_graph_ptr->following(lanelet)) {
+    if (
+      planner_data.preferred_lanelets.front().id() != lanelet.id() &&
+      exists(planner_data.route_lanelets, lanelet)) {
+      return lanelet;
```

Comment:
Returning the first lanelet in the next lanelets is not good as a specification of the utility function.
Let's change the function to `get_next_lanelets_within_route`, and the node side will use the first one.

Reply:
The next lanelet **in a route** should be single.
So, please add the error log like
```
RCLCPP_WARN(rclcpp::get_logger("path_generator.utils"), "The multiple next lanelets in a route are found not as expected. Internal calculation might have failed.");
```
when there are multiple next lanelets.

Returning the first lanelet in this case is fine.